drv/lpc55-sprot-server/src/main.rs

```diff
@@ -175,280 +114,618 @@ fn main() -> ! {
-        IoStatus::Flush,
-        &server.io.rx_buf,
-        bytes_read,
-        &mut server.io.tx_buf,
-        server.state,
-    );
+        gpio,
+        state: State::init(),
+        stats: IoStats::default(),
+    }
+}
 
-    loop {
-        server.io.pio(transmit);
-        let iostat = if server.io.rxcount == 0
-            && !server.io.underrun
-            && !server.io.overrun
-        {
-            IoStatus::Flush
-        } else {
-            IoStatus::IOResult {
-                underrun: server.io.underrun,
-                overrun: server.io.overrun,
-            }
-        };
-        transmit = match server.handler.handle(
-            transmit, // true if previous loop transmitted.
-            iostat,
-            &server.io.rx_buf,
-            server.io.rxcount,
-            &mut server.io.tx_buf,
-            server.state,
-        ) {
-            Some(txlen) => {
-                ringbuf_entry!(Trace::HandlerReturnSize(txlen.0));
-                true
-            }
-            None => {
-                ringbuf_entry!(Trace::HandleReturnNone);
-                false
-            }
+// Container for spi and gpio
+struct Io {
+    spi: crate::spi_core::Spi,
+    gpio: drv_lpc55_gpio_api::Pins,
+    state: State,
+    stats: IoStats,
+}
+
+/// The state of the IO state machine
+///
+/// Logically, the sprot protocol operates in a request/response fashion,
+/// despite the fact that bytes must simultaneously be clocked in and out during SPI
+/// transactions.
+///
+/// As part of our logical state machine, the RoT can only be reading a request from
+/// the SP or writing a reply. There can only be one request in flight at a time.
+/// When desynchronization is detected it must be corrected.
+///
+/// The RoT always starts in State::Read(ReadState::WaitingForRequest);
+#[derive(Clone, Copy, PartialEq)]
+pub enum State {
+    Read(ReadState),
+    Write(WriteState),
+}
+
+impl State {
+    /// Initialize the state Io state machine
+    fn init() -> State {
+        State::Read(ReadState::WaitingForRequest)
+    }
+
+    /// Return the read state. Panics if the RoT is not currently reading.
+    fn read(&self) -> ReadState {
+        match self {
+            State::Read(s) => *s,
+            _ => panic!(),
         }
     }
+
+    /// Return the write state. Panics if the RoT is not currently writing.
+    fn write(&self) -> WriteState {
+        match self {
+            State::Write(s) => *s,
+            _ => panic!(),
+        }
+    }
+}
+
+/// The state of the IO state machine when trying to read a request from the SP
+#[derive(Clone, Copy, PartialEq)]
+pub enum ReadState {
+    /// The RoT is waiting for CSn to be asserted
+    WaitingForRequest,
+
+    /// The RoT is in frame and reading data until it sees CSn de-asserted
+    InFrame,
+
+    /// The frame was read cleanly.
+    ///
+    /// In this case we were:
+    ///   1. In `WaitingForRequest` and saw CSn asserted and CSn de-asserted at the
```

Comment:
Need to fix this comment. We don't check for CSn de-asserted in `WaitingForRequest`